In [5]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb
import requests
import io

%matplotlib inline

#### Question 1
What's the version of NumPy that you installed?

In [6]:
np.__version__

'1.20.3'

#### Getting the data

In [30]:
# Using requests library
url = 'https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv'
response = requests.get(url).content

In [71]:
# Reading the dataset
df=  pd.read_csv(io.StringIO(response.decode('utf-8')))

In [72]:
# An overview of the dataset
df.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

#### Question 2
How many records are in the dataset?

Here you need to specify the number of rows

In [29]:
df.describe()

Year    Engine HP  Engine Cylinders  Number of Doors  \
count  11914.000000  11845.00000      11884.000000     11908.000000   
mean    2010.384338    249.38607          5.628829         3.436093   
std        7.579740    109.19187          1.780559         0.881315   
min     1990.000000     55.00000          0.000000         2.000000   
25%     2007.000000    170.00000          4.000000         2.000000   
50%     2015.000000    227.00000          6.000000         4.000000   
75%     2016.000000    300.00000          6.000000         4.000000   
max     2017.000000   1001.00000         16.000000         4.000000   

        highway MPG      city mpg    Popularity          MSRP  
count  11914.000000  11914.000000  11914.000000  1.191400e+04  
mean      26.637485     19.733255   1554.911197  4.059474e+04  
std        8.863001      8.987798   1441.855347  6.010910e+04  
min       12.000000      7.000000      2.000000  2.000000e+03  
25%       22.000000     16.000000    549.000000  2.100000e+04  
50%       26.000000     18.000000   1385.000000  2.999500e+04  
75%       30.000000     22.000000   2009.000000  4.223125e+04  
max      354.000000    137.000000   5657.000000  2.065902e+06

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Make               11914 non-null  object 
 1   Model              11914 non-null  object 
 2   Year               11914 non-null  int64  
 3   Engine Fuel Type   11911 non-null  object 
 4   Engine HP          11845 non-null  float64
 5   Engine Cylinders   11884 non-null  float64
 6   Transmission Type  11914 non-null  object 
 7   Driven_Wheels      11914 non-null  object 
 8   Number of Doors    11908 non-null  float64
 9   Market Category    8172 non-null   object 
 10  Vehicle Size       11914 non-null  object 
 11  Vehicle Style      11914 non-null  object 
 12  highway MPG        11914 non-null  int64  
 13  city mpg           11914 non-null  int64  
 14  Popularity         11914 non-null  int64  
 15  MSRP               11914 non-null  int64  
dtypes: float64(3), int64(5

 - From above, RangeIndex gives **11914**

### Question 3
Who are the most frequent car manufacturers (top-3) according to the dataset?

In [13]:
df.Make.value_counts()

Chevrolet        1123
Ford              881
Volkswagen        809
Toyota            746
Dodge             626
Nissan            558
GMC               515
Honda             449
Mazda             423
Cadillac          397
Mercedes-Benz     353
Suzuki            351
BMW               334
Infiniti          330
Audi              328
Hyundai           303
Volvo             281
Subaru            256
Acura             252
Kia               231
Mitsubishi        213
Lexus             202
Buick             196
Chrysler          187
Pontiac           186
Lincoln           164
Oldsmobile        150
Land Rover        143
Porsche           136
Saab              111
Aston Martin       93
Plymouth           82
Bentley            74
Ferrari            69
FIAT               62
Scion              60
Maserati           58
Lamborghini        52
Rolls-Royce        31
Lotus              29
Tesla              18
HUMMER             17
Maybach            16
Alfa Romeo          5
McLaren             5
Spyker    

- Chevrolet (1123), Ford (881) and Volkswagen (809) are top 3 based on the value_counts

#### Question 4
What's the number of unique Audi car models in the dataset?

In [14]:
# Getting unique items
df.loc[df['Make'] == 'Audi']['Model'].nunique()

34

- 34 Audi models are unique

#### Question 5
How many columns in the dataset have missing values?

In [15]:
df.isnull().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

- Answer is  **5**.
- From above, by manually counting, we have 5 columns have integers other than zero (0), i.e., Engine Fuel Type 3, Engine HP 69, Engine Cylinders 30, Number of Doors 6 and Market Category 3742 have missing values.
- Hence, answer is **5**

#### Question 6: Does the median value change after filling missing values?
1. Find the median value of "Engine Cylinders" column in the dataset.
2. Next, calculate the most frequent value of the same "Engine Cylinders".
3. Use the fillna method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.
4. Now, calculate the median value of "Engine Cylinders" once again.

- Has it changed?


In [50]:
median_engine_cylinders = df['Engine Cylinders'].median()
median_engine_cylinders

6.0

In [51]:
mode_engine_cylinders = df['Engine Cylinders'].mode()
mode_engine_cylinders

0    4.0
dtype: float64

In [73]:
# Copyin the dataset
clean_cylinders = df['Engine Cylinders']

In [90]:
# Checking empty rows in the column Engine Cylinders
clean_cylinders.isna().sum()

30

In [97]:
clean_cylinders.fillna(mode_engine_cylinders, inplace=True)

In [98]:
clean_cylinders.median()

6.0

- No change, the median is the same, i.e.,  **6.0**

#### Questions 7: Value of the first element of w*

- Select all the "Lotus" cars from the dataset.
- Select only columns "Engine HP", "Engine Cylinders".
- Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).
- Get the underlying NumPy array. Let's call it X.
- Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
- Invert XTX.
- Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].
- Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
- What's the value of the first element of w?


In [100]:
df_lotus = df[df['Make'] == 'Lotus']
df_lotus.sample(5)

Make  Model  Year                Engine Fuel Type  Engine HP  \
3916  Lotus  Elise  2010     premium unleaded (required)      218.0   
4293  Lotus  Exige  2009  premium unleaded (recommended)      240.0   
4294  Lotus  Exige  2010  premium unleaded (recommended)      240.0   
4260  Lotus  Evora  2011     premium unleaded (required)      276.0   
4267  Lotus  Evora  2014     premium unleaded (required)      345.0   

      Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
3916               4.0            MANUAL  rear wheel drive              2.0   
4293               4.0            MANUAL  rear wheel drive              2.0   
4294               4.0            MANUAL  rear wheel drive              2.0   
4260               6.0            MANUAL  rear wheel drive              2.0   
4267               6.0            MANUAL  rear wheel drive              2.0   

              Market Category Vehicle Size Vehicle Style  highway MPG  \
3916  Exotic,High-Performance      Compact   Convertible           26   
4293  Exotic,High-Performance      Compact         Coupe           26   
4294  Exotic,High-Performance      Compact         Coupe           26   
4260  Exotic,High-Performance      Compact         Coupe           26   
4267  Exotic,High-Performance      Compact         Coupe           26   

      city mpg  Popularity   MSRP  
3916        20         613  54990  
4293        20         613  65690  
4294        20         613  65690  
4260        18         613  65500  
4267        17         613  79980

In [101]:
lotus_df_columns = df_lotus[['Engine HP', 'Engine Cylinders']]
lotus_df_columns.head()

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3914      189.0               4.0
3915      189.0               4.0
3916      218.0               4.0

In [102]:
lotus_df_columns.duplicated().sum()

20

In [103]:
# Replacing with inplace=True
lotus_df_columns.drop_duplicates(inplace=True)

C:\Users\User\Anaconda3\lib\site-packages\pandas\util\_decorators.py:311: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return func(*args, **kwargs)


In [104]:
# Checking to confirm is changes are effective
lotus_df_columns.duplicated().sum()

0

In [105]:
# Checking rows, should give 9 based on instructions
lotus_df_columns.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9 entries, 3912 to 4293
Data columns (total 2 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Engine HP         9 non-null      float64
 1   Engine Cylinders  9 non-null      float64
dtypes: float64(2)
memory usage: 216.0 bytes


In [106]:
# Get the underlying NumPy array. Let's call it X.
X = lotus_df_columns.to_numpy()

In [107]:
# Displaying X
X

array([[189.,   4.],
       [218.,   4.],
       [217.,   4.],
       [350.,   8.],
       [400.,   6.],
       [276.,   6.],
       [345.,   6.],
       [257.,   4.],
       [240.,   4.]])

In [108]:
# Compute matrix-matrix multiplication between the transpose of X and X.
# To get the transpose, use X.T. Let's call the result XTX.
XT = X.T
XT


array([[189., 218., 217., 350., 400., 276., 345., 257., 240.],
       [  4.,   4.,   4.,   8.,   6.,   6.,   6.,   4.,   4.]])

In [109]:
XTX = np.dot(XT, X)
XTX

array([[7.31684e+05, 1.34100e+04],
       [1.34100e+04, 2.52000e+02]])

In [115]:
# Inverting XTX
XTX_Inverse = np.linalg.inv(XTX)
XTX_Inverse

array([[ 5.53084235e-05, -2.94319825e-03],
       [-2.94319825e-03,  1.60588447e-01]])

In [116]:
y = np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800])

In [117]:
w = np.dot(np.dot(XTX_Inverse, XT), y)

In [118]:
# Displaying w
w

array([  4.59494481, -63.56432501])

- Hence, from above, the first value in **w** is **4.59494481**

#### Saving Dataframe to csv for future use

In [114]:
df.to_csv('data_data.csv')